d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px; height: 163px">
</div>

# CSE Coding Assignment
## Instructions

- Please answer all questions
- You can use any language you wish (e.g. Python, Scala, SQL...)
- Several Markdown cells require completion. Please edit the Markdown cells to include your answer.
- Your final notebook should compile without errors when you click "Run All"

**Please do not publish questions. This is a confidential assignment.**

### Creating a Cluster

You will need to create a Databricks Cluster. More information on this process is available here: https://docs.databricks.com/user-guide/clusters/create.html

## Getting Started

**REQUIRED:** Run the following cells exactly as written to retrieve the necessary Coding Assignment Data Sets from Amazon S3.

In [0]:
%sh curl --remote-name-all 'https://files.training.databricks.com/assessments/cse-take-home/{covertype,kafka,treecover,u.data,u.item}.csv'


[1/5]: https://files.training.databricks.com/assessments/cse-take-home/covertype.csv --> covertype.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   133  100   133    0     0   1176      0 --:--:-- --:--:-- --:--:--  1176

[2/5]: https://files.training.databricks.com/assessments/cse-take-home/kafka.csv --> kafka.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  135k  100  135k    0     0  2119k      0 --:--:-- --:--:-- --:--:-- 2119k

[3/5]: https://files.training.databricks.com/assessments/cse-take-home/treecover.csv --> treecover.csv
  % Total    % Received % Xferd  Average Speed   Time    Time 

In [0]:
dbutils.fs.cp("file:/databricks/driver/covertype.csv", "dbfs:/FileStore/tmp/covertype.csv")
dbutils.fs.cp("file:/databricks/driver/kafka.csv", "dbfs:/FileStore/tmp/kafka.csv")
dbutils.fs.cp("file:/databricks/driver/treecover.csv", "dbfs:/FileStore/tmp/treecover.csv")
dbutils.fs.cp("file:/databricks/driver/u.data.csv", "dbfs:/FileStore/tmp/u.data.csv")
dbutils.fs.cp("file:/databricks/driver/u.item.csv", "dbfs:/FileStore/tmp/u.item.csv")

Out[2]: True

## Part 1: Reading and Parsing Data

### Question 1:  Code Challenge - Load a CSV

- Load the CSV file at `dbfs:/FileStore/tmp/nl/treecover.csv` into a DataFrame.
- Use Apache Spark to read in the data, assigned to the variable `treeCoverDF`.
- Your method to get the CSV file into Databricks isn't graded. We are only concerned with how you use Spark to parse and load the actual data. 
- Please use the `inferSchema` option.

In [0]:
path = "/FileStore/tmp/treecover.csv"

treeCoverDF=spark.read.option("header", "true").option("inferSchema" , "true").csv(path)
treeCoverDF.show()

+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
| Id|Elevation|Aspect|Slope|Horizontal_Distance_To_Hydrology|Vertical_Distance_To_Hydrology|Horizontal_Distance_To_Roadways|Horizontal_Distance_To_Fire_Points|Cover_Type|Soil_Type|Wilderness_Area|Hillshade|
+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
|  1|     2596|    51|    3|                             258|                             0|                            510|                              6279|         5|       29|              1|      9am|
|  2|     2590|    56|    2|                             212|                            -6|                            390|                              6225|         5|  

### Question 2:  Code Challenge - Print the Schema

Use Apache Spark to display the Schema of the `treeCoverDF` Dataframe.

In [0]:
treeCoverDF.printSchema()

root
 |-- Id: integer (nullable = true)
 |-- Elevation: integer (nullable = true)
 |-- Aspect: integer (nullable = true)
 |-- Slope: integer (nullable = true)
 |-- Horizontal_Distance_To_Hydrology: integer (nullable = true)
 |-- Vertical_Distance_To_Hydrology: integer (nullable = true)
 |-- Horizontal_Distance_To_Roadways: integer (nullable = true)
 |-- Horizontal_Distance_To_Fire_Points: integer (nullable = true)
 |-- Cover_Type: integer (nullable = true)
 |-- Soil_Type: integer (nullable = true)
 |-- Wilderness_Area: integer (nullable = true)
 |-- Hillshade: string (nullable = true)



### Question 3:  Code Challenge - Rows & Columns

Use Apache Spark to display the number of rows and columns in the DataFrame.

In [0]:
treeCoverDF.agg({'Elevation': 'count'}).show()
print("Number of rows in treeCoverDF: ", treeCoverDF.count())
print("Number of Columns in treeCoverDF: ", len(treeCoverDF.columns))

+----------------+
|count(Elevation)|
+----------------+
|           15120|
+----------------+

Number of rows in treeCoverDF:  15120
Number of Columns in treeCoverDF:  12


#Part 2: Analysis

### Question 4:  Code Challenge - Summary Statistics for a Feature

Use Apache Spark to answer these questions about the `treeCoverDF` DataFrame:
- What is the range - minimum and maximum - of values for the feature `elevation`?
- What are the mean and standard deviation of the feature `elevation`?

In [0]:
from pyspark.sql.functions import max,min # max/min method from pyspark.sql.functions resolved the issue of max/min value of column as it overrides built in python methods
treeCoverDF.agg(max(treeCoverDF.Elevation)).show()#1
treeCoverDF.agg(min(treeCoverDF.Elevation)).show()#2
#3treeCoverDF.agg({'Elevation': 'min'}).show()
#4treeCoverDF.agg({'Elevation': 'max'}).show()
#5treeCoverDF.select([min("Elevation")]).show()
#6treeCoverDF.select([max("Elevation")]).show()
#7treeCoverDF.registerTempTable("treeCoverDF_tbl")
#8spark.sql("SELECT MIN(Elevation) as minval_Of_Elevation FROM treeCoverDF_tbl").show()
#9spark.sql("SELECT MAX(Elevation) as maxval_Of_Elevation FROM treeCoverDF_tbl").show()

from pyspark.sql.functions import mean , stddev , col

df_stats = treeCoverDF.select(
    mean(col('Elevation')).alias('_mean'),
    stddev(col('Elevation')).alias('_std')
).collect()
print(df_stats[0]['_mean'],df_stats[0]['_std'])

+--------------+
|max(Elevation)|
+--------------+
|          3849|
+--------------+

+--------------+
|min(Elevation)|
+--------------+
|          1863|
+--------------+

2749.3225529100528 417.67818734804985


In [0]:
treeCoverDF.select([min("Elevation")]).show()
treeCoverDF.select([max("Elevation")]).show()

+--------------+
|min(Elevation)|
+--------------+
|          1863|
+--------------+

+--------------+
|max(Elevation)|
+--------------+
|          3849|
+--------------+



In [0]:
treeCoverDF.agg({'Elevation': 'min'}).show()
treeCoverDF.agg({'Elevation': 'max'}).show()

+--------------+
|min(Elevation)|
+--------------+
|          1863|
+--------------+

+--------------+
|max(Elevation)|
+--------------+
|          3849|
+--------------+



In [0]:
treeCoverDF.registerTempTable("treeCoverDF_tbl")
spark.sql("SELECT MIN(Elevation) as minval_Of_Elevation FROM treeCoverDF_tbl").show()
spark.sql("SELECT MAX(Elevation) as maxval_Of_Elevation FROM treeCoverDF_tbl").show()

/databricks/spark/python/pyspark/sql/dataframe.py:146: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn(
+-------------------+
|minval_Of_Elevation|
+-------------------+
|               1863|
+-------------------+

+-------------------+
|maxval_Of_Elevation|
+-------------------+
|               3849|
+-------------------+



### Answer #4:

- Min `elevation`: `1863`
- Max `elevation`: `3849`
- Mean `elevation`: `2749.3225529100528`
- Standard Deviation of `elevation`: `417.67818734804985`

### Question 5:  Code Challenge - Record Count

Use Apache Spark to answer the following question:
- How many entries in the dataset have an `elevation` greater than or equal to 2749.32 meters **AND** a `Cover_Type` of 1 or 2?

In [0]:
#Be cautious abt '(' and ')' while using bitwise operators | / &
dfl=treeCoverDF.filter(((treeCoverDF["Cover_Type"] == 1) | (treeCoverDF["Cover_Type"] == 2)) & (treeCoverDF["Elevation"]>=2749.32)).count()
print(dfl)

3883


### Question 6: Code Challenge - Compute a Percentage

Use Apache Spark to answer the following question:
- What percentage of entries with `Cover_Type` 1 or 2 have an `elevation` at or above 2749.32 meters?

In [0]:
percent= dfl/treeCoverDF.count()*100
print(percent,"%")

25.681216931216934 %


### Question 7: Code Challenge - Visualize Feature Distribution

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart that helps visualize the distribution of different Wilderness Areas in our dataset

In [0]:
from pyspark.sql.functions import *

Wilderness_Aray = treeCoverDF.groupBy(treeCoverDF["Wilderness_Area"]).count()
display(Wilderness_Aray)

Wilderness_Area count 1 3597 3 6349 4 4675 2 499

### Question 8: Code Challenge - Visualize Average Elevation by Cover Type 

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart showing the average elevation of each cover type with string labels for cover type

**NOTE: you will need to match the integer values in the column `treeCoverDF.Cover_Type` to the string values in `dbfs:/FileStore/tmp/nl/covertype.csv` to retrieve the Cover Type Labels. It is recommended to use an Apache Spark join.**

In [0]:
from pyspark.sql.functions import *
fft1=treeCoverDF.groupby(treeCoverDF["Cover_Type"]).mean().select("Cover_Type","avg(Elevation)")
path = "/FileStore/tmp/covertype.csv"
CovertypeDF=spark.read.option("header", "true").option("inferSchema" , "true").csv(path)
fft2=fft1.join(CovertypeDF).where(fft1["Cover_Type"]==CovertypeDF["cover_type_key"]).select("cover_type_label","avg(Elevation)")
display(fft2)

cover_type_label avg(Elevation) Spruce/Fir 3128.025925925926 Douglas-fir 2423.276851851852 Ponderosa Pine 2398.4231481481484 Aspen 2786.801388888889 Cottonwood/Willow 2223.42037037037 Krummholz 3362.7699074074076 Lodgepole Pine 2922.5402777777776

#Part 3: Data Ingestion, Cleansing, and Transformations

## Instructions 

This is a multi-step, data pipeline question in which you need to achieve a few objectives to build a successful job.

### Data Sets

#### `u.data.csv`

- The full u data set, 100000 ratings by 943 users on 1682 items. 
- Each user has rated at least 20 movies.  
- Users and items are numbered consecutively from 1. 
- The data is randomly ordered. 
- This is a tab separated file consisting of four columns: 
   - user id 
   - movie id 
   - rating 
   - date (unix seconds since 1/1/1970 UTC)

#### Desired schema

- `user_id INTEGER`
- `movie_id INTEGER`
- `rating INTEGER`
- `date DATE `

#### `u.item.csv`

- This is a `|` separated file consisting of six columns:
   - movie id
   - movie title
   - release date
   - video release date
   - IMDb URL
   - genre
- movie ids in this file match movie ids in `u.data`.

#### Desired schema

- `movie_id INTEGER`
- `movie_title STRING`

### Question 9:  Code Challenge - Load DataFrames

Use Apache Spark to perform the following:
1. define the correct schemas for each Data Set to be imported as described above  
   **note:** 
      - for `u.data.csv`, `date` *must* be stored using `DateType` with the format `yyyy-MM-dd`
      - you may need to ingest `timestamp` data using `IntegerType`
      - be sure to drop unneccesary columns for `u.item.csv`
1. import the two files as DataFrames names `uDataDF` and `uItemDF` using the schemas you defined and these paths:
   - `dbfs:/FileStore/tmp/u.data.csv`
   - `dbfs:/FileStore/tmp/u.item.csv`
1. order the `uDataDF` DataFrame by the `date` column

**NOTE:** Please display the DataFrames, `uDataDF` and `uItemDF` after loading.

#### `uDataDF`

In [0]:
from pyspark.sql.functions import *
from pyspark.sql.types import *
from datetime import datetime

path= "/FileStore/tmp/u.data.csv"
schema = StructType([StructField("User_id", IntegerType(), True),
                     StructField("Movie_id", IntegerType(), True),
                     StructField("Rating", IntegerType(), True),
                     StructField("date_time", IntegerType()),
                    ])
DF1=spark.read.option("delimiter", "\t").csv(path,schema=schema)
DF2=DF1.withColumn("date_time_fmt",from_unixtime(DF1.date_time))\
.select("User_id","Movie_id","Rating",date_format('date_time_fmt','yyyy-mm-dd').alias('Date'),date_format('date_time_fmt','hh:mm:ss').alias('Time'))
uDataDF=DF2.orderBy(["Date"],ascending=False)
#uDataDF=uDataDF.withColumn("Date",to_date("Date"))
uDataDF.display()

#DF1 = DF1.withColumn("date", DF1["date_time"].cast(TimestampType())) # Here typecasting is done on date_time to covert from unix to timestamp formats
#1uDataDF=pyspark.sql.functions.split(DF1['date_time'], ' ') ###Here split can be done but to obtain two different data frames is being created in #1,2,3
#2df_date = df.withColumn('Date', split_col.getItem(0))
#3df_time = df.withColumn('timestamp', split_col.getItem(1))


User_id Movie_id Rating Date Time 650 135 4 1998-59-31 09:59:05 7 609 3 1998-59-31 01:59:09 766 357 4 1998-59-31 01:59:18 454 984 3 1998-59-31 08:59:28 90 657 5 1998-59-31 10:59:50 7 537 3 1998-59-31 01:59:09 766 616 3 1998-59-31 01:59:49 766 82 3 1998-59-31 01:59:18 766 527 5 1998-59-31 01:59:18 650 519 4 1998-59-31 09:59:05 90 568 5 1998-59-31 10:59:25 650 216 4 1998-59-31 09:59:06 650 604 3 1998-59-31 10:59:38 7 509 5 1998-59-31 01:59:38 618 815 4 1998-59-31 01:59:12 938 117 3 1998-59-31 02:59:10 938 275 4 1998-59-31 02:59:10 650 313 4 1998-59-31 09:59:06 659 367 3 1998-59-31 10:59:26 90 942 4 1998-59-31 10:59:25 90 151 2 1998-59-31 10:59:50 659 186 3 1998-59-31 10:59:57 650 187 2 1998-59-31 09:59:45 650 191 4 1998-59-31 09:59:06 938 508 4 1998-59-31 02:59:27 90 836 5 1998-59-31 10:59:50 766 526 2 1998-59-31 01:59:18 26 323 2 1998-59-31 12:59:44 650 144 3 1998-59-31 09:59:45 650 272 4 1998-59-31 09:59:06 650 385 4 1998-59-31 09:59:45 938 100 5 1998-59-31 02:59:10 650 210 3 1998-59-31 09:59:45 650 118 4 1998-59-31 09:59:06 650 89 4 1998-59-31 09:59:45 938 248 1 1998-59-31 02:59:50 90 220 4 1998-59-31 10:59:25 938 257 5 1998-59-31 02:59:10 766 664 2 1998-59-31 01:59:49 7 151 4 1998-59-31 01:59:09 650 198 4 1998-59-31 09:59:06 650 168 4 1998-59-31 09:59:06 618 458 3 1998-59-31 01:59:39 7 188 5 1998-59-31 01:59:38 90 515 5 1998-59-31 10:59:25 766 234 4 1998-59-31 01:59:18 90 162 5 1998-59-31 10:59:50 7 668 4 1998-59-31 01:59:38 650 133 4 1998-59-31 09:59:06 938 181 5 1998-59-31 02:59:50 650 82 3 1998-59-31 09:59:45 7 217 4 1998-59-31 01:59:38 720 258 4 1998-59-30 12:59:22 625 357 3 1998-59-30 12:59:44 720 995 4 1998-59-30 12:59:22 720 345 2 1998-59-30 12:59:22 358 1266 4 1998-59-30 02:59:04 625 257 4 1998-59-30 03:59:03 625 484 4 1998-59-30 12:59:43 720 310 4 1998-59-30 12:59:22 585 543 3 1998-59-30 06:59:53 932 225 2 1998-59-30 09:59:45 263 234 4 1998-59-30 10:59:52 932 1454 4 1998-59-30 09:59:45 585 1501 4 1998-59-30 06:59:53 720 321 4 1998-59-30 12:59:22 625 50 5 1998-59-30 03:59:03 720 272 4 1998-59-30 12:59:22 585 30 4 1998-59-30 06:59:53 932 414 4 1998-59-30 09:59:19 585 213 5 1998-59-30 06:59:53 263 210 3 1998-59-30 10:59:52 263 205 5 1998-59-30 10:59:52 263 144 4 1998-59-30 10:59:52 358 666 3 1998-59-30 02:59:52 263 614 3 1998-59-30 10:59:52 278 752 5 1998-59-30 09:59:24 278 538 4 1998-59-30 09:59:24 585 116 3 1998-59-30 06:59:53 625 222 4 1998-59-30 03:59:03 720 872 3 1998-59-30 12:59:40 660 167 2 1998-59-29 07:59:25 660 926 2 1998-59-29 07:59:47 660 50 4 1998-59-29 06:59:40 660 380 2 1998-59-29 07:59:47 840 213 4 1998-59-29 08:59:59 480 443 4 1998-59-29 09:59:06 840 180 5 1998-59-29 08:59:03 660 809 2 1998-59-29 07:59:25 660 393 2 1998-59-29 07:59:01 660 181 4 1998-59-29 06:59:58 840 637 3 1998-59-29 08:59:59 934 213 4 1998-59-29 04:59:04 934 99 3 1998-59-29 05:59:39 660 550 2 1998-59-29 07:59:01 660 121 2 1998-59-29 06:59:14 480 234 4 1998-59-29 09:59:29 934 170 4 1998-59-29 04:59:04 117 1014 3 1998-59-28 08:59:52 328 1313 4 1998-59-28 04:59:09 495 679 3 1998-59-28 02:59:44 495 1119 4 1998-59-28 02:59:44 537 697 2 1998-59-28 11:59:26 537 549 2 1998-59-28 11:59:25 610 12 5 1998-59-28 09:59:17 495 161 4 1998-59-28 02:59:06 452 729 1 1998-59-28 09:59:34 610 1 4 1998-59-28 09:59:17 610 185 5 1998-59-28 09:59:51 505 328 4 1998-59-28 01:59:35 328 905 3 1998-59-28 04:59:59 610 607 5 1998-59-28 09:59:17 328 349 2 1998-59-28 04:59:09 610 176 4 1998-59-28 09:59:17 747 304 4 1998-59-28 03:59:30 537 72 1 1998-59-28 11:59:26 401 237 3 1998-59-27 08:59:27 339 269 5 1998-59-27 08:59:39 401 248 3 1998-59-27 08:59:27 339 157 4 1998-59-27 08:59:39 339 428 5 1998-59-27 08:59:09 401 1011 3 1998-59-27 08:59:27 401 515 4 1998-59-27 08:59:27 548 14 1 1998-59-27 11:59:42 339 488 5 1998-59-27 08:59:39 591 100 5 1998-59-27 10:59:25 559 502 4 1998-59-27 09:59:06 835 325 5 1998-59-27 08:59:51 774 254 1 1998-59-27 05:59:04 794 24 5 1998-59-27 09:59:17 828 10 3 1998-59-27 09:59:30 591 285 5 1998-59-27 10:59:25 931 362 3 1998-59-27 09:59:30 313 89

#### `uItemDF`

In [0]:
from pyspark.sql.functions import *
from pyspark.sql.types import *

path= "/FileStore/tmp/u.item.csv"
schemat = StructType([StructField("Movie_id", IntegerType(), True),
                     StructField("Movie_title", StringType(), True)
                    ])
uItemDF=spark.read.option("delimiter", "|").csv(path,schema=schemat)
uItemDF.display()

Movie_id Movie_title 1 Toy Story (1995) 2 GoldenEye (1995) 3 Four Rooms (1995) 4 Get Shorty (1995) 5 Copycat (1995) 6 Shanghai Triad (Yao a yao yao dao waipo qiao) (1995) 7 Twelve Monkeys (1995) 8 Babe (1995) 9 Dead Man Walking (1995) 10 Richard III (1995) 11 Seven (Se7en) (1995) 12 Usual Suspects, The (1995) 13 Mighty Aphrodite (1995) 14 Postino, Il (1994) 15 Mr. Holland's Opus (1995) 16 French Twist (Gazon maudit) (1995) 17 From Dusk Till Dawn (1996) 18 White Balloon, The (1995) 19 Antonia's Line (1995) 20 Angels and Insects (1995) 21 Muppet Treasure Island (1996) 22 Braveheart (1995) 23 Taxi Driver (1976) 24 Rumble in the Bronx (1995) 25 Birdcage, The (1996) 26 Brothers McMullen, The (1995) 27 Bad Boys (1995) 28 Apollo 13 (1995) 29 Batman Forever (1995) 30 Belle de jour (1967) 31 Crimson Tide (1995) 32 Crumb (1994) 33 Desperado (1995) 34 Doom Generation, The (1995) 35 Free Willy 2: The Adventure Home (1995) 36 Mad Love (1995) 37 Nadja (1994) 38 Net, The (1995) 39 Strange Days (1995) 40 To Wong Foo, Thanks for Everything! Julie Newmar (1995) 41 Billy Madison (1995) 42 Clerks (1994) 43 Disclosure (1994) 44 Dolores Claiborne (1994) 45 Eat Drink Man Woman (1994) 46 Exotica (1994) 47 Ed Wood (1994) 48 Hoop Dreams (1994) 49 I.Q. (1994) 50 Star Wars (1977) 51 Legends of the Fall (1994) 52 Madness of King George, The (1994) 53 Natural Born Killers (1994) 54 Outbreak (1995) 55 Professional, The (1994) 56 Pulp Fiction (1994) 57 Priest (1994) 58 Quiz Show (1994) 59 Three Colors: Red (1994) 60 Three Colors: Blue (1993) 61 Three Colors: White (1994) 62 Stargate (1994) 63 Santa Clause, The (1994) 64 Shawshank Redemption, The (1994) 65 What's Eating Gilbert Grape (1993) 66 While You Were Sleeping (1995) 67 Ace Ventura: Pet Detective (1994) 68 Crow, The (1994) 69 Forrest Gump (1994) 70 Four Weddings and a Funeral (1994) 71 Lion King, The (1994) 72 Mask, The (1994) 73 Maverick (1994) 74 Faster Pussycat! Kill! Kill! (1965) 75 Brother Minister: The Assassination of Malcolm X (1994) 76 Carlito's Way (1993) 77 Firm, The (1993) 78 Free Willy (1993) 79 Fugitive, The (1993) 80 Hot Shots! Part Deux (1993) 81 Hudsucker Proxy, The (1994) 82 Jurassic Park (1993) 83 Much Ado About Nothing (1993) 84 Robert A. Heinlein's The Puppet Masters (1994) 85 Ref, The (1994) 86 Remains of the Day, The (1993) 87 Searching for Bobby Fischer (1993) 88 Sleepless in Seattle (1993) 89 Blade Runner (1982) 90 So I Married an Axe Murderer (1993) 91 Nightmare Before Christmas, The (1993) 92 True Romance (1993) 93 Welcome to the Dollhouse (1995) 94 Home Alone (1990) 95 Aladdin (1992) 96 Terminator 2: Judgment Day (1991) 97 Dances with Wolves (1990) 98 Silence of the Lambs, The (1991) 99 Snow White and the Seven Dwarfs (1937) 100 Fargo (1996) 101 Heavy Metal (1981) 102 Aristocats, The (1970) 103 All Dogs Go to Heaven 2 (1996) 104 Theodore Rex (1995) 105 Sgt. Bilko (1996) 106 Diabolique (1996) 107 Moll Flanders (1996) 108 Kids in the Hall: Brain Candy (1996) 109 Mystery Science Theater 3000: The Movie (1996) 110 Operation Dumbo Drop (1995) 111 Truth About Cats & Dogs, The (1996) 112 Flipper (1996) 113 Horseman on the Roof, The (Hussard sur le toit, Le) (1995) 114 Wallace & Gromit: The Best of Aardman Animation (1996) 115 Haunted World of Edward D. Wood Jr., The (1995) 116 Cold Comfort Farm (1995) 117 Rock, The (1996) 118 Twister (1996) 119 Maya Lin: A Strong Clear Vision (1994) 120 Striptease (1996) 121 Independence Day (ID4) (1996) 122 Cable Guy, The (1996) 123 Frighteners, The (1996) 124 Lone Star (1996) 125 Phenomenon (1996) 126 Spitfire Grill, The (1996) 127 Godfather, The (1972) 128 Supercop (1992) 129 Bound (1996) 130 Kansas City (1996) 131 Breakfast at Tiffany's (1961) 132 Wizard of Oz, The (1939) 133 Gone with the Wind (1939) 134 Citizen Kane (1941) 135 2001: A Space Odyssey (1968) 136 Mr. Smith Goes to Washington (1939) 137 Big Night (1996) 138 D3: The Mighty Ducks (1996) 139 Love Bug, The (1969) 140 Homeward Bound: The Incredible Journey (1993) 141 20,000 Leagues Und

### Question 10:  Code Challenge - Perform a Join

Use Apache Spark to do the following:
- join `uDataDF` and `uItemDf` on `movie_id` as a new DataFrame called `uMovieDF`  
   **note:** make sure you do not create duplicate `movie_id` columns
   
**NOTE:** Please display the DataFrame `uMovieDF`.

In [0]:
uMovieDF=uDataDF.join(uItemDF).where(uDataDF["Movie_id"]==uItemDF["Movie_id"])
uMovieDF.display()

User_id Movie_id Rating Date Time Movie_id Movie_title 196 242 3 1997-55-04 03:55:49 242 Kolya (1996) 186 302 3 1998-22-04 07:22:22 302 L.A. Confidential (1997) 22 377 1 1997-18-07 07:18:36 377 Heavyweights (1994) 244 51 2 1997-02-27 05:02:03 51 Legends of the Fall (1994) 166 346 1 1998-33-02 05:33:16 346 Jackie Brown (1997) 298 474 4 1998-20-07 02:20:06 474 Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb (1963) 115 265 2 1997-51-03 05:51:28 265 Hunt for Red October, The (1990) 253 465 5 1998-34-03 06:34:27 465 Jungle Book, The (1994) 305 451 3 1998-20-01 09:20:17 451 Grease (1978) 6 86 3 1997-16-31 09:16:53 86 Remains of the Day, The (1993) 62 257 2 1997-07-12 10:07:14 257 Men in Black (1997) 286 1014 5 1997-38-17 03:38:45 1014 Romy and Michele's High School Reunion (1997) 200 222 5 1997-05-05 09:05:40 222 Star Trek: First Contact (1996) 210 40 3 1998-59-27 09:59:54 40 To Wong Foo, Thanks for Everything! Julie Newmar (1995) 224 29 3 1998-40-21 11:40:57 29 Batman Forever (1995) 303 785 3 1997-28-14 05:28:38 785 Only You (1994) 122 387 5 1997-47-11 05:47:39 387 Age of Innocence, The (1993) 194 274 2 1997-36-14 08:36:34 274 Sabrina (1995) 291 1042 4 1997-42-21 09:42:24 1042 Just Cause (1995) 234 1184 2 1998-47-08 11:47:17 1184 Endless Summer 2, The (1994) 119 392 4 1998-13-30 04:13:34 392 Man Without a Face, The (1993) 167 486 4 1998-54-16 02:54:12 486 Sabrina (1954) 299 144 4 1997-55-26 03:55:20 144 Die Hard (1988) 291 118 2 1997-24-21 09:24:38 118 Twister (1996) 308 1 4 1998-28-17 05:28:52 1 Toy Story (1995) 95 546 2 1997-16-10 09:16:06 546 Broken Arrow (1996) 38 95 5 1998-14-13 01:14:54 95 Aladdin (1992) 102 768 2 1998-40-02 01:40:50 768 Casper (1995) 63 277 4 1997-10-01 11:10:01 277 Restoration (1995) 160 234 5 1997-33-14 08:33:05 234 Jaws (1975) 50 246 3 1997-38-17 01:38:49 246 Chasing Amy (1997) 301 98 4 1997-03-14 05:03:47 98 Silence of the Lambs, The (1991) 225 193 4 1997-35-14 08:35:27 193 Right Stuff, The (1983) 290 88 4 1997-46-28 03:46:03 88 Sleepless in Seattle (1993) 97 194 3 1998-54-08 05:54:20 194 Sting, The (1973) 157 274 4 1998-33-07 10:33:55 274 Sabrina (1995) 181 1081 1 1997-17-08 04:17:03 1081 Curdled (1996) 278 603 5 1998-02-30 10:02:10 603 Rear Window (1954) 276 796 1 1997-45-20 09:45:32 796 Speechless (1994) 7 32 4 1998-28-31 01:28:52 32 Crumb (1994) 10 16 4 1997-01-26 06:01:17 16 French Twist (Gazon maudit) (1995) 284 304 4 1998-48-20 08:48:42 304 Fly Away Home (1996) 201 979 2 1998-17-06 07:17:13 979 Trigger Effect, The (1996) 276 564 3 1997-43-20 09:43:25 564 Tales from the Hood (1995) 287 327 5 1997-18-27 04:18:36 327 Cop Land (1997) 246 201 5 1998-33-16 03:33:14 201 Evil Dead II (1987) 242 1137 5 1997-33-17 04:33:16 1137 Beautiful Thing (1996) 249 241 5 1997-46-16 12:46:34 241 Last of the Mohicans, The (1992) 99 4 5 1998-18-03 03:18:17 4 Get Shorty (1995) 178 332 3 1997-43-22 08:43:57 332 Kiss the Girls (1997) 251 100 4 1998-38-31 06:38:04 100 Fargo (1996) 81 432 2 1997-58-11 01:58:51 432 Fantasia (1940) 260 322 4 1998-08-23 02:08:18 322 Murder at 1600 (1997) 25 181 5 1998-23-26 10:23:35 181 Return of the Jedi (1983) 59 196 5 1998-38-23 03:38:08 196 Dead Poets Society (1989) 72 679 2 1997-46-20 02:46:04 679 Conan the Barbarian (1981) 87 384 4 1997-18-18 06:18:47 384 Naked Gun 33 1/3: The Final Insult (1994) 290 143 5 1997-11-25 04:11:33 143 Sound of Music, The (1965) 42 423 5 1997-08-03 12:08:07 423 E.T. the Extra-Terrestrial (1982) 292 515 4 1997-06-02 11:06:17 515 Boot, Das (1981) 115 20 3 1997-43-03 05:43:29 20 Angels and Insects (1995) 20 288 1 1997-06-16 08:06:24 288 Scream (1996) 201 219 4 1998-51-06 06:51:13 219 Nightmare on Elm Street, A (1984) 13 526 3 1997-10-14 11:10:53 526 Ben-Hur (1959) 246 919 4 1998-22-16 03:22:29 919 City of Lost Children, The (1995) 138 26 5 1997-23-08 09:23:52 26 Brothers McMullen, The (1995) 167 232 1 1998-52-16 02:52:21 232 Young Guns (1988) 60 427 5 1997-30-28 04:30:20 427 To Kill a Mockingbird (1962) 57 304 5 1998-49-01 11:49:41 304 Fly Away

### Question 11:  Code Challenge - Perform an Aggregation

Use Apache Spark to do the following:
1. create an aggregate DataFrame, `aggDF` by
  1. extracting the year from the `date` (of the review)
  1. getting the average rating of each film per year as a column named `average_rating`
  1. ordering descending by year and average rating
1. write the resulting dataframe to a table named "movie_by_year_average_rating" in the Default database  
   **note:** use `mode(overwrite)` 

#### Desired Schema
The schema of you resulting DataFrame should be:
- `year INTEGER`
- `movie_title STRING`
- `average_rating DOUBLE`

**NOTE:** Please display the DataFrame `aggDF`.

In [0]:
from pyspark.sql.functions import *

DF1 = uMovieDF.withColumn("Year", split(col("Date"), "-").getItem(0)).select("Year","Movie_title","Rating").groupby("Year","Movie_title").mean()
DF2=DF1.select("Year","Movie_title",col("avg(Rating)").alias("average_rating"))
aggDF=DF2.orderBy(["Year","average_rating"],ascending=False)
aggDF.display(truncate=False)
aggDF.printSchema()
aggDF.write.csv("/FileStore/tables/aggDF/")


spark.sql("Drop TABLE default.movie_by_year_average_rating")
sqlContext.sql("create table default.movie_by_year_average_rating(Year INT, Movie_title STRING, average_rating Float) ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' LINES TERMINATED BY '\n'")
spark.sql("LOAD DATA INPATH '/FileStore/tables/aggDF/' OVERWRITE INTO TABLE default.movie_by_year_average_rating")
#spark.sql("select * from default.movie_by_year_average_rating  where average_rating is Null").show()
#uMovieDF.select("Date","Movie_title","Rating").where(uMovieDF["Movie_title"]=="Spy Hard (1996)").show()
#from pyspark.sql import SQLContext
#sqlContext = new org.apache.spark.sql.SQLContext(sc)


Year Movie_title average_rating 1998 Star Kid (1997) 5.0 1998 Prefontaine (1997) 5.0 1998 Great Day in Harlem, A (1994) 5.0 1998 Entertaining Angels: The Dorothy Day Story (1996) 5.0 1998 The Deadly Cure (1996) 5.0 1998 Visitors, The (Visiteurs, Les) (1993) 5.0 1998 Faust (1994) 5.0 1998 Marlene Dietrich: Shadow and Light (1996) 5.0 1998 Mina Tannenbaum (1994) 5.0 1998 Year of the Horse (1997) 4.75 1998 Pather Panchali (1955) 4.666666666666667 1998 Safe (1995) 4.6 1998 8 Seconds (1994) 4.5 1998 Crossfire (1947) 4.5 1998 Some Mother's Son (1996) 4.5 1998 Innocents, The (1961) 4.5 1998 Anna (1996) 4.5 1998 Faithful (1996) 4.5 1998 Everest (1998) 4.5 1998 Hotel de Love (1996) 4.5 1998 Wallace & Gromit: The Best of Aardman Animation (1996) 4.5 1998 Grateful Dead (1995) 4.5 1998 Shall We Dance? (1996) 4.454545454545454 1998 Casablanca (1942) 4.452830188679245 1998 Usual Suspects, The (1995) 4.411214953271028 1998 Close Shave, A (1995) 4.409090909090909 1998 12 Angry Men (1957) 4.407407407407407 1998 Schindler's List (1993) 4.3893129770992365 1998 Star Wars (1977) 4.379032258064516 1998 Rear Window (1954) 4.358695652173913 1998 One Flew Over the Cuckoo's Nest (1975) 4.352941176470588 1998 Maybe, Maybe Not (Bewegte Mann, Der) (1994) 4.333333333333333 1998 Whole Wide World, The (1996) 4.333333333333333 1998 Letter From Death Row, A (1998) 4.333333333333333 1998 Maya Lin: A Strong Clear Vision (1994) 4.333333333333333 1998 Bitter Sugar (Azucar Amargo) (1996) 4.333333333333333 1998 Stonewall (1995) 4.333333333333333 1998 To Kill a Mockingbird (1962) 4.3232323232323235 1998 Secrets & Lies (1996) 4.318840579710145 1998 North by Northwest (1959) 4.311688311688312 1998 Wrong Trousers, The (1993) 4.3 1998 Godfather, The (1972) 4.2897727272727275 1998 Shawshank Redemption, The (1994) 4.26605504587156 1998 Good Will Hunting (1997) 4.262886597938144 1998 Vertigo (1958) 4.259740259740259 1998 Wild Reeds (1994) 4.25 1998 Margaret's Museum (1995) 4.25 1998 Cement Garden, The (1993) 4.25 1998 Citizen Kane (1941) 4.244186046511628 1998 Third Man, The (1949) 4.243243243243243 1998 Titanic (1997) 4.24223602484472 1998 L.A. Confidential (1997) 4.224852071005917 1998 Maltese Falcon, The (1941) 4.2153846153846155 1998 African Queen, The (1951) 4.214285714285714 1998 Waiting for Guffman (1996) 4.2105263157894735 1998 Raise the Red Lantern (1991) 4.205882352941177 1998 Mr. Smith Goes to Washington (1939) 4.204545454545454 1998 Manchurian Candidate, The (1962) 4.203125 1998 Killing Fields, The (1984) 4.2 1998 39 Steps, The (1935) 4.2 1998 As Good As It Gets (1997) 4.196428571428571 1998 Silence of the Lambs, The (1991) 4.193939393939394 1998 Godfather: Part II, The (1974) 4.19 1998 Boot, Das (1981) 4.183908045977011 1998 Sunset Blvd. (1950) 4.181818181818182 1998 Lone Star (1996) 4.176470588235294 1998 His Girl Friday (1940) 4.172413793103448 1998 Chinatown (1974) 4.171875 1998 Blade Runner (1982) 4.169354838709677 1998 Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb (1963) 4.168539325842697 1998 Paths of Glory (1957) 4.166666666666667 1998 Amadeus (1984) 4.163793103448276 1998 Empire Strikes Back, The (1980) 4.162650602409639 1998 Cinema Paradiso (1988) 4.160714285714286 1998 Raiders of the Lost Ark (1981) 4.15934065934066 1998 Great Escape, The (1963) 4.158730158730159 1998 Sling Blade (1996) 4.153846153846154 1998 Lawrence of Arabia (1962) 4.151898734177215 1998 Lost Horizon (1937) 4.15 1998 Manon of the Spring (Manon des sources) (1986) 4.147058823529412 1998 Eat Drink Man Woman (1994) 4.146341463414634 1998 In the Bleak Midwinter (1995) 4.142857142857143 1998 Raging Bull (1980) 4.140350877192983 1998 Some Folks Call It a Sling Blade (1993) 4.133333333333334 1998 Bridge on the River Kwai, The (1957) 4.133333333333334 1998 Fresh (1994) 4.125 1998 Garden of Finzi-Contini, The (Giardino dei Finzi-Contini, Il) (1970) 4.111111111111111 1998 Ran (1985) 4.111111111111111 1998 Apt Pupil (1998) 4.1 1998 Princess Bride, The (1987) 4.1 1998 Ph

root
 |-- Year: string (nullable = true)
 |-- Movie_title: string (nullable = true)
 |-- average_rating: double (nullable = true)

+----+--------------------+--------------+
|Year|         Movie_title|average_rating|
+----+--------------------+--------------+
|1998|           "Visitors|          null|
|1998|"Great Day in Harlem|          null|
|1998|          "Innocents|          null|
|1998|     "Usual Suspects|          null|
|1998|        "Close Shave|          null|
|1998|   "Whole Wide World|          null|
|1998|"Letter From Deat...|          null|
|1998|              "Maybe|          null|
|1998|     "Wrong Trousers|          null|
|1998|          "Godfather|          null|
|1998|"Shawshank Redemp...|          null|
|1998|      "Cement Garden|          null|
|1998|          "Third Man|          null|
|1998|     "Maltese Falcon|          null|
|1998|      "African Queen|          null|
|1998|"Manchurian Candi...|          null|
|1998|     "Killing Fields|          null|
|1998|   

## Part 4: Fun with JSON

JSON values are typically passed by message brokers such as Kafka or Kinesis in a string encoding. When consumed by a Spark Structured Streaming application, this json must be converted into a nested object in order to be used.

Below is a list of json strings that represents how data might be passed from a message broker.

**Note:** Make sure to run the cell below to retrieve the sample data.

In [0]:
%python


sampleJson = [
 ('{"user":100, "ips" : ["191.168.192.101", "191.168.192.103", "191.168.192.96", "191.168.192.99"]}',), 
 ('{"user":101, "ips" : ["191.168.192.102", "191.168.192.105", "191.168.192.103", "191.168.192.107"]}',), 
 ('{"user":102, "ips" : ["191.168.192.105", "191.168.192.101", "191.168.192.105", "191.168.192.107"]}',), 
 ('{"user":103, "ips" : ["191.168.192.96", "191.168.192.100", "191.168.192.107", "191.168.192.101"]}',), 
 ('{"user":104, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.102", "191.168.192.99"]}',), 
 ('{"user":105, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.100", "191.168.192.96"]}',), 
]

### Question 12:  Code Challenge - Count the IPs

Use any coding techniques known to you to parse this list of JSON strings to answer the following question:
- how many occurrences of each IP address are in this list?

#### Desired Output
Your results should be this:


| ip | count |
|:-:|:-:|
| `191.168.192.96` | `3` |
| `191.168.192.99` | `6` |
| `191.168.192.100` | `2` |
| `191.168.192.101` | `3` |
| `191.168.192.102` | `2` |
| `191.168.192.103` | `2` |
| `191.168.192.105` | `3` |
| `191.168.192.107` | `3` |

**NOTE:** The order of your results is not important.

In [0]:
from pyspark.sql.functions import *
from pyspark.sql.types import *
from pyspark.sql import Row

list2=[]
for i in sampleJson:
  x=list(i)
  y=x[0].split(': ["')
  for j in y:
    z=j.split('", "')
    for k in z:
      a=k.split('"]')
      for l in a:
        if len(l)==15 or len(l)==14:
          list2.append(l)
rdd1 = sc.parallelize(list2)
row_rdd = rdd1.map(lambda x: Row(x))
df_ip=sqlContext.createDataFrame(row_rdd,['ip']) 
df_ip.select("ip").groupBy("ip").count().orderBy("ip",ascending=True).display()
### below code also works however the column name argument is not possible to assign
#001>>>df_ip=spark.createDataFrame(list2, StringType())
#002>>>print(df_ip)
#003>>>df_ip.select("value").groupBy("value").count().display()

##### pandas created dataframe didnt recognize select ,groupby
#101,102
#101>>>df_ip = pd.DataFrame({'ip':list2}) 
#102>>>df_ip.select("ip").groupBy("ip").count().show()

###### tried creating sql context didnt work as it was considered as object instead of row format
##error:'TypeError: StructType can not accept object '191.168.192.105' in type <class 'str'>'.
#201,202,203
#201>>>schema = StructType([StructField("ip", StringType(), True)])
#202>>>rdd_ip = sc.parallelize(list2)
#203>>>df_ip = sqlContext.createDataFrame(rdd_ip, schema)

ip count 191.168.192.100 2 191.168.192.101 3 191.168.192.102 2 191.168.192.103 2 191.168.192.105 3 191.168.192.107 3 191.168.192.96 3 191.168.192.99 6

DataFrame[ip: string]


## Part 5: The Databricks API

### Question 13: Conceptual Question - the Databricks API

In 4-5 sentences, please explain what the Databricks API is used for at a high-level.

### Answer:

- The Databricks API is to connect to a single node cluster in community edition
- We have the dbfs as file system and databases
- We can also mount the external filesystem on to Databricks
- we can run Spark SQL, Pyspark, Python on notebooks
- This is industrial wide used platform for big data processing

### Question 14: Conceptual Question - Explain an API Call

In 4-5 sentences, please explain what this API call. Be sure to discuss some key attributes about the cluster.

```
$ curl -n -X POST -H 'Content-Type: application/json'                      \
  -d '{                                                                     \
  "cluster_name": "high-concurrency-cluster",                               \
  "spark_version": "4.2.x-scala2.11",                                       \
  "node_type_id": "i3.xlarge",                                              \
  "spark_conf":{                                                            \
        "spark.databricks.cluster.profile":"serverless",                    \
        "spark.databricks.repl.allowedLanguages":"sql,python,r"             \
     },                                                                     \
     "aws_attributes":{                                                     \
        "zone_id":"us-west-2c",                                             \
        "first_on_demand":1,                                                \
        "availability":"SPOT_WITH_FALLBACK",                                \
        "spot_bid_price_percent":100                                        \
     },                                                                     \
   "custom_tags":{                                                          \
        "ResourceClass":"Serverless"                                        \
     },                                                                     \
       "autoscale":{                                                        \
        "min_workers":1,                                                    \
        "max_workers":2                                                     \
     },                                                                     \
  "autotermination_minutes":10                                              \
}' https://dogfood.staging.cloud.databricks.com/api/2.0/clusters/create '
```

### Answer:

- This is to create cluster with AWS serverless lambda resource
- Associated with programing languages set to execute SQL,Python,R 
- Autoscaling feature of min 1 node max 2 nodes based on 1 node as on demand instance and other as spot instance with 100% bid price
- Also auto termination of idle node with 10 sec heartbeat rate

## Part 6: Security

### Question 15: Conceptual Question - Security on Databricks

Using the Databricks Documentation, what would you recommend to a Databricks and AWS customer for **securely** storing and accessing their data.

### Answer:

- It is good practice to use the AWS S3 to store the data as storage 
- Mount the S3 root folder to dbfs path 
- This makes the efficent model for fast computation along side with storage

# This is the end of the official test. Bonus below!

## Part 7: Bonus: Data Science & Machine Learning

### Question 16: Conceptual Question - A Skewed Feature

One of these lines is the *mean* of this feature. The other is the *median*. Which of these lines is the **mean** - the red line or the black line?

<img width=400px src=https://www.evernote.com/l/AAEycL6CQ0hLi5V5pIo91Ko-Pfk2i0AnGyMB/image.png>

### Answer:

- Black line is the mean

### Question 17: Conceptual Question - Exploratory Data Analysis

The plots below show the distribution of home selling prices differentiated by a few categorical features. Based on these plots, **which of these categorical features** - Property Type, Exterior Quality, or Month Sold - would you expect to be most associated with Price? **Why**?

<img width=1600px src=https://www.evernote.com/l/AAHulkcc20hHSJV6D1udKiwSDCN0S6oV_5YB/image.png>

### Answer:

- Not sure how to read the plots

### Question 18: Conceptual Question - Analyze Model Performance

Consider the following results for a decision tree model against training and testing data sets:

`decision tree regression - train r2 score: 0.9944`  
`decision tree regression - test r2 score:  0.3119`


What is your assessment of this model?

### Answer:

- Not sure what is expected to be answered

### Question 19: Conceptual Question - Model Selection

A series of models has been built using the same training data, but each with a subset of features.

Consider the following results for a series of logistic regression models and then answer this question:

- Which model would you choose and why?
- What other things would you want to look at and why?

| model number | feature subset | logistic regression test accuracy|
|:-:|:-:|:-:|
| 1| feat_1 |	 0.631|
| 2| feat_2 |	 0.552|
| 3| feat_3 |	 0.868|
| 4| feat_4 |	 0.868|
| 5| feat_1, feat_2 |	 0.657|
| 6| feat_1, feat_3 |	 0.947|
| 7| feat_1, feat_4 |	 0.921|
| 8| feat_2, feat_3 |	 0.947|
| 9| feat_2, feat_4 |	 0.973|
| 10| feat_3, feat_4 |	 0.947|
| 11| feat_1, feat_2, feat_3 |	 0.947|
| 12| feat_1, feat_2, feat_4 |	 0.947|
| 13| feat_1, feat_3, feat_4 |	 0.947|
| 14| feat_2, feat_3, feat_4 |	 0.973|
| 15| feat_1, feat_2, feat_3, feat_4 |	 0.973|

### Answer:

- Model 9 is the best, as the regression test accuracy is high and lesser in feature subset
- Model 14 as its second best

-sandbox
&copy; 2019 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>